source/Nuke.Common/Execution/DockerExecutor.cs

```diff
@@ -0,0 +1,166 @@
+
+namespace Nuke.Common.Execution
+{
+    internal class DockerExecutor
+    {
+        public static bool IsDockerStep(NukeBuild build, ExecutableTarget target)
+        {
+            var attribute = GetRunInDockerContainerAttribute(build, target);
+            return attribute != null;
+        }
+
+        [CanBeNull]
+        private static RunInDockerContainerAttribute GetRunInDockerContainerAttribute(NukeBuild nukeBuild, ExecutableTarget target)
+        {
+            var targetProperty = ExecutableTargetFactory.GetTargetProperties(nukeBuild.GetType()).FirstOrDefault(x => x.Name == target.Name);
+            return targetProperty?.GetCustomAttribute<RunInDockerContainerAttribute>();
+        }
+
+        public static void Execute(NukeBuild build, ExecutableTarget target)
+        {
+            var attribute = GetRunInDockerContainerAttribute(build, target)!;
+
+            PullImageIfRequired(attribute);
+
+            PublishBuildProject(attribute);
+
+            CopyPackageReferences();
+
+            var workingDirectory = attribute.Platform == PlatformFamily.Windows ? "c:\\Build" : "/build";
+
+            var path = attribute.Platform == PlatformFamily.Windows
+                ? $"c:\\build\\{NukeBuild.RootDirectory.GetWinRelativePathTo(NukeBuild.TemporaryDirectory)}\\nukebuild\\{Path.GetFileName(NukeBuild.BuildAssembly)}"
+                : $"/build/{NukeBuild.RootDirectory.GetUnixRelativePathTo(NukeBuild.TemporaryDirectory)}/nukebuild/{Path.GetFileName(NukeBuild.BuildAssembly)}";
+
+            var args = new[]
+                       {
+                           path,
+                           "--nologo",
+                           "--skip",
+                           "--host", "NukeInDocker",
+                           "--target", $"\"{target.Name}\"",
+                           "--root", workingDirectory
+                       };
+
+            var userProfile = EnvironmentInfo.SpecialFolder(SpecialFolders.UserProfile);
+            var tempFile = GetEnvFile(workingDirectory, userProfile);
+
+            try
+            {
+                var volumes = new []
+                              {
+                                  $"{NukeBuild.RootDirectory}:{workingDirectory}"//,
+                                  //$"{userProfile}:{userProfile}"
+                              };
+                DockerTasks.DockerRun(settings => settings
+                    .EnableRm()
+                    .SetImage(attribute.Image)
+                    .SetVolume(volumes)
+                    .SetCommand("dotnet")
+                    .SetWorkdir(workingDirectory)
+                    .SetEnvFile(tempFile)
+                    .SetArgs(args.Concat(attribute.Args)));
+            }
+            finally
+            {
+                //File.Delete(tempFile);
+            }
+        }
+
+        private static void CopyPackageReferences()
+        {
+            Log.Information("Inlining PackageDownload references from build project");
```

Comment:
Need to copy the packages into the container, as they're not included in the published output.
There's probably a better way.